In [4]:
import pandas as pd
import numpy as np
import glob
import json
%matplotlib inline

In [5]:
path = 'psyreg_output/'
filenames = glob.glob(path + '*.json')

In [6]:
len(filenames)

7131

### Ein Level

In [6]:
df = pd.DataFrame()
for filename in filenames:
    with open(filename) as json_data:
        data = json.load(json_data)
    row = pd.Series({
        'idPerson': data['idPerson'],
        'name': data['name'],
        'vorname': data['vorname']
    })
    df = df.append(row, ignore_index=True)

In [7]:
df.head(5)

idPerson               name    vorname
0   11036.0  Schweizer-Vüllers    Andreas
1     495.0       Altwegg-Artz    Sibylle
2    2719.0             Prchal      Alice
3   14978.0        Grob Müller   Gabriela
4    4452.0              Borer  Christine

'psyreg_output/11036_psyreg.json'

In [4]:
data = []
with open(filenames[0]) as json_data:
    data = json.load(json_data)
data

{'idPerson': 11036,
 'name': 'Schweizer-Vüllers',
 'vorname': 'Andreas',
 'geschlechtId': 12000,
 'geburtsdatum': None,
 'jahrgang': 1946,
 'sprachkompetenzIds': [13000],
 'nationalitaet1Id': 10074,
 'nationalitaet2Id': None,
 'gln': '7601007633722',
 'uid': None,
 'adressenBewilligungen': [{'kantonId': 11020,
   'praxisname': '-',
   'strasse': 'Gemeindestrasse 27',
   'plz': '8032',
   'ort': 'Zürich',
   'telefon1': '+41 44 252 00 37',
   'emailadresse': 'an-schweizer@bluewin.ch',
   'zusatz1': None,
   'zusatz2': ''}],
 'adressenNeunzigTageDienstleister': [],
 'bewilligungenErteilt': [{'kantonId': 11020,
   'auflageEinschraenkungen': [{'typId': 8003,
     'datumEntscheid': '1999-09-23T00:00:00',
     'datumBefristung': '2022-07-23T00:00:00'}],
   'datumEntscheid': '1999-09-23T00:00:00',
   'datumErteilung': '2018-10-23T15:26:48.997',
   'datumBefristung': '2019-06-28T13:53:35.347'}],
 'bewilligungenEntzogen': [],
 'bewilligungenVerweigert': [],
 'bewilligungenKeineBewilligung': [],

In [39]:
a = flattenDict(data)
a

[('idPerson', 11036),
 ('name', 'Schweizer-Vüllers'),
 ('vorname', 'Andreas'),
 ('geschlechtId', 12000),
 ('geburtsdatum', None),
 ('jahrgang', 1946),
 ('sprachkompetenzIds', [13000]),
 ('nationalitaet1Id', 10074),
 ('nationalitaet2Id', None),
 ('gln', '7601007633722'),
 ('uid', None),
 ('adressenBewilligungen',
  [{'kantonId': 11020,
    'praxisname': '-',
    'strasse': 'Gemeindestrasse 27',
    'plz': '8032',
    'ort': 'Zürich',
    'telefon1': '+41 44 252 00 37',
    'emailadresse': 'an-schweizer@bluewin.ch',
    'zusatz1': None,
    'zusatz2': ''}]),
 ('adressenNeunzigTageDienstleister', []),
 ('bewilligungenErteilt',
  [{'kantonId': 11020,
    'auflageEinschraenkungen': [{'typId': 8003,
      'datumEntscheid': '1999-09-23T00:00:00',
      'datumBefristung': '2022-07-23T00:00:00'}],
    'datumEntscheid': '1999-09-23T00:00:00',
    'datumErteilung': '2018-10-23T15:26:48.997',
    'datumBefristung': '2019-06-28T13:53:35.347'}]),
 ('bewilligungenEntzogen', []),
 ('bewilligungenVerwe

In [8]:
len(df)

7131

Es sind offenbar alle Einträge da.

Also der Versuch jetzt mehr Zeilen einzulesen:


In [7]:
import pandas as pd
import numpy as np
import glob
import json
%matplotlib inline

path = 'psyreg_output/'
filenames = glob.glob(path + '*.json')

df = pd.DataFrame()
for file in filenames:
    with open(file) as json_data:
        data = json.load(json_data)
        minidict = {
                    'id': data['idPerson'],
                    'name': data['name'],
                    'vorname': data['vorname'],
                    'geschlecht': data['geschlechtId'],
                    'jahrgang': data['jahrgang'],
                    'sprache': data['sprachkompetenzIds'],
                    'nationalität_1': data['nationalitaet1Id'],
                    'nationalität_2': data['nationalitaet2Id']
        }
        abschluesse = []
        for abschluss in data['abschluesse']:
            abschluss_dict = {'abschluss': abschluss['abschlussId'],
                    'abschluss_date': abschluss['ausstellungsdatum'],
                    'ausstellungsland': abschluss['ausstellungslandId'],
                    'anerkennungstyp': abschluss['anerkennungstypId']}
            abschluesse.append(abschluss_dict)
        for tätig_in in data['adressenBewilligungen']
                taetig_dict ={
                    'plz': tätig_in['plz'],
                    'ort': tätig_in['ort'],
                    'entzogen': data['bewilligungenEntzogen'],
                    'verweigert': data['bewilligungenVerweigert'],
                    'keine' : data['bewilligungenKeineBewilligung']

                })
                df = df.append(row, ignore_index=True)


In [8]:
len(df)

5604

In [4]:
df.head(10)

abschluss       abschluss_date anerkennungstyp ausstellungsland entzogen  \
0      None                 None            None             None       []   
1     55004  2002-10-01T00:00:00            None            10001       []   
2     55001  1979-01-01T00:00:00            None            10001       []   
3     55004  1990-06-16T00:00:00            None            10001       []   
4     55004                 None            None            10001       []   
5     55001  2000-01-01T00:00:00            None            10001       []   
6     55004  2006-04-01T00:00:00            None            10001       []   
7     55001  1992-01-01T00:00:00            None            10001       []   
8     55004  2000-09-16T00:00:00            None            10001       []   
9     55001  1985-01-01T00:00:00            None            10001       []   

   geschlecht       id jahrgang keine               name  nationalität_1  \
0     12000.0  11036.0     1946    []  Schweizer-Vüllers         10074.0   
1     12001.0   4452.0     1954    []              Borer         10001.0   
2     12000.0   2249.0     1951    []           Limacher         10001.0   
3     12000.0   2249.0     1951    []           Limacher         10001.0   
4     12001.0   4102.0     1946    []    Pingler-Maassen         10074.0   
5     12001.0   2686.0     1973    []     Peternac-Pfund         10001.0   
6     12001.0   2686.0     1973    []     Peternac-Pfund         10001.0   
7     12000.0   3094.0     1962    []            Schmutz         10001.0   
8     12000.0   3094.0     1962    []            Schmutz         10001.0   
9     12001.0   1373.0     1959    []         Frei-Käser         10001.0   

  nationalität_2           ort   plz  sprache verweigert    vorname  
0           None        Zürich  8032  [13000]         []    Andreas  
1           None        Zürich  8057  [13000]         []  Christine  
2           None        Zürich  8032  [13000]         []   Bernhard  
3           None        Zürich  8032  [13000]         []   Bernhard  
4           None         Aarau  5000  [13000]         []       Gila  
5           None  Schaffhausen  8200  [13000]         []  Franziska  
6           None  Schaffhausen  8200  [13000]         []  Franziska  
7           None        Zürich  8053  [13000]         []   Matthias  
8           None        Zürich  8053  [13000]         []   Matthias  
9           None      Bettlach  2544  [13000]         []      Anita

Das df braucht noch einige Anpassungen. Aber zuerst mal ein erster Check. 

In [33]:
len(df)

5604

Etwas stimmt nicht. Das df müsste mehr Reihen haben, da 7131 Datensätze vorhanden sind, und Leute mit mehr Bewilligungen oder mehr Abschlüssen zusätzliche Reihen generieren. 
Ein Versuch, den Code anzupassen. 

In [17]:
import pandas as pd
import numpy as np
import glob
import json
%matplotlib inline

path = 'psyreg_output/'
filenames = glob.glob(path + '*.json')

df = pd.DataFrame()
for file in filenames:
    with open(file) as json_data:
        data = json.load(json_data)
        
        row = pd.Series({
            'id': data['idPerson'],
            'name': data['name'],
            'vorname': data['vorname'],
            'geschlecht': data['geschlechtId'],
            'jahrgang': data['jahrgang'],
            'sprache': data['sprachkompetenzIds'],
            'nationalität_1': data['nationalitaet1Id'],
            'nationalität_2': data['nationalitaet2Id'],

        for abschluss in data['abschluesse']:
            'abschluss': abschluss['abschlussId'],
            'abschluss_date': abschluss['ausstellungsdatum'],
            'ausstellungsland': abschluss['ausstellungslandId'],
            'anerkennungstyp': abschluss['anerkennungstypId'],
        
        for tätig_in in data['adressenBewilligungen']:
            'plz': tätig_in['plz'],
            'ort': tätig_in['ort'],
            'entzogen': data['bewilligungenEntzogen'],
            'verweigert': data['bewilligungenVerweigert'],
            'keine' : data['bewilligungenKeineBewilligung']
            })
        
        df = df.append(row, ignore_index=True)

SyntaxError: invalid syntax (<ipython-input-17-6e57dc2b9527>, line 25)

In [36]:
df['ausstellungsland'].unique()

array([None, 10001, 10002, 10032, 10052, 10069, 10082, 10077, 10059,
       10062, 10058, 10076, 10075, 10055, 10085, 10066, 10005, 10064,
       10023], dtype=object)

In [37]:
df['ausstellungsland'].value_counts()

10001    5114
10002     182
10052      23
10032      22
10085       4
10062       3
10069       2
10066       2
10075       2
10076       2
10077       2
10064       1
10082       1
10058       1
10055       1
10023       1
10005       1
10059       1
Name: ausstellungsland, dtype: int64

In [42]:
df['nationalität_1'].value_counts()

10001.0    4612
10002.0     396
10074.0     374
13000.0      53
10052.0      45
10032.0      27
10077.0      13
10049.0       8
10025.0       7
10023.0       7
10075.0       7
10066.0       6
10064.0       6
10044.0       4
10014.0       4
10069.0       4
10055.0       4
10059.0       3
10192.0       3
10026.0       3
10005.0       2
10016.0       2
10013.0       2
10085.0       2
10033.0       2
10050.0       2
10058.0       2
10045.0       2
10007.0       1
10063.0       1
Name: nationalität_1, dtype: int64

In [43]:
df['nationalität_2'].value_counts()

10002    49
10032    23
10077     9
10001     7
10072     6
10044     4
10007     4
10026     4
10037     4
10076     4
10085     3
10049     2
10052     2
10033     2
10060     2
10023     2
10017     2
10013     2
10010     2
10063     2
10098     2
10056     1
10022     1
Name: nationalität_2, dtype: int64

In [5]:
df['anerkennungstyp'].value_counts()

27018    133
27021     52
27041     25
27040     19
27042      8
27039      5
Name: anerkennungstyp, dtype: int64

# So macht es Plotti - die Daten flachmachen (flatten)


In [7]:
data = []
with open(filenames[0]) as json_data:
    data = json.load(json_data)
data

{'idPerson': 11036,
 'name': 'Schweizer-Vüllers',
 'vorname': 'Andreas',
 'geschlechtId': 12000,
 'geburtsdatum': None,
 'jahrgang': 1946,
 'sprachkompetenzIds': [13000],
 'nationalitaet1Id': 10074,
 'nationalitaet2Id': None,
 'gln': '7601007633722',
 'uid': None,
 'adressenBewilligungen': [{'kantonId': 11020,
   'praxisname': '-',
   'strasse': 'Gemeindestrasse 27',
   'plz': '8032',
   'ort': 'Zürich',
   'telefon1': '+41 44 252 00 37',
   'emailadresse': 'an-schweizer@bluewin.ch',
   'zusatz1': None,
   'zusatz2': ''}],
 'adressenNeunzigTageDienstleister': [],
 'bewilligungenErteilt': [{'kantonId': 11020,
   'auflageEinschraenkungen': [{'typId': 8003,
     'datumEntscheid': '1999-09-23T00:00:00',
     'datumBefristung': '2022-07-23T00:00:00'}],
   'datumEntscheid': '1999-09-23T00:00:00',
   'datumErteilung': '2018-10-23T15:26:48.997',
   'datumBefristung': '2019-06-28T13:53:35.347'}],
 'bewilligungenEntzogen': [],
 'bewilligungenVerweigert': [],
 'bewilligungenKeineBewilligung': [],

Test von Codefragment. 
https://stackoverflow.com/questions/6027558/flatten-nested-dictionaries-compressing-keys

In [5]:
# Python3 code to iterate over a list 
list = [1, 3, 5, 7, 9] 
   
# Using enumerate()  
for i, val in enumerate(list): 
    print (i, ",",val) 

0 , 1
1 , 3
2 , 5
3 , 7
4 , 9


Das Problem mit den Daten ist, dass sie mehrfach verschachtelt sind. Das kostet auch Plotti einige Testrunden mit Code ab. 

In [8]:
data

{'idPerson': 11036,
 'name': 'Schweizer-Vüllers',
 'vorname': 'Andreas',
 'geschlechtId': 12000,
 'geburtsdatum': None,
 'jahrgang': 1946,
 'sprachkompetenzIds': [13000],
 'nationalitaet1Id': 10074,
 'nationalitaet2Id': None,
 'gln': '7601007633722',
 'uid': None,
 'adressenBewilligungen': [{'kantonId': 11020,
   'praxisname': '-',
   'strasse': 'Gemeindestrasse 27',
   'plz': '8032',
   'ort': 'Zürich',
   'telefon1': '+41 44 252 00 37',
   'emailadresse': 'an-schweizer@bluewin.ch',
   'zusatz1': None,
   'zusatz2': ''}],
 'adressenNeunzigTageDienstleister': [],
 'bewilligungenErteilt': [{'kantonId': 11020,
   'auflageEinschraenkungen': [{'typId': 8003,
     'datumEntscheid': '1999-09-23T00:00:00',
     'datumBefristung': '2022-07-23T00:00:00'}],
   'datumEntscheid': '1999-09-23T00:00:00',
   'datumErteilung': '2018-10-23T15:26:48.997',
   'datumBefristung': '2019-06-28T13:53:35.347'}],
 'bewilligungenEntzogen': [],
 'bewilligungenVerweigert': [],
 'bewilligungenKeineBewilligung': [],

Eine der verschachtelten Listen von Dicts ist zum Beispiel das Feld 'bewilligungErteilt'. 

In [9]:
type(data["bewilligungenErteilt"]) == type([])

True

In [22]:
output = {}
for k,v in data.items(): #k für Key, v = value #.items holt k und v
    if type(v) == type([]):
        dummy_keys = []
        for i in range(0,3):
            try:
                element = v[i]
                if len(dummy_keys) > 0:
                    dummy_keys = []
                if type(element) == int:
                    print("been here int")
                    output[k] = element
                else:
                    print("been here")
                    for k_inside, v_inside in element.items():
                        dummy_keys.append(k_inside)
                        print(dummy_keys)
                        if type(v_inside) == type([]):
                            for inner_element in v_inside:
                                for k_inside_inside,v_inside_inside in inner_element.items():
                                    dummy_keys.append(k_inside+"-"+k_inside_inside)
                                    print(dummy_keys)
                                    output[k+"-"+str(i)+"-"+k_inside+"-"+k_inside_inside] = v_inside_inside
                        else:
                            output[k+"-"+str(i)+"-"+k_inside] = v_inside
            except:
                print(dummy_keys)
                for key in dummy_keys:
                    output[k+"-"+str(i)+"-"+ key] = ''
    else:
        print("outer")
        output[k] = v
pd.DataFrame(output,index=[0]).columns

outer
outer
outer
outer
outer
outer
been here int
[]
[]
outer
outer
outer
outer
been here
['kantonId']
['kantonId', 'praxisname']
['kantonId', 'praxisname', 'strasse']
['kantonId', 'praxisname', 'strasse', 'plz']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1', 'zusatz2']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1', 'zusatz2']
['kantonId', 'praxisname', 'strasse', 'plz', 'ort', 'telefon1', 'emailadresse', 'zusatz1', 'zusatz2']
[]
[]
[]
been here
['kantonId']
['kantonId', 'auflageEinschraenkungen']
['kantonId', 'auflageEinschraenkungen', 'auflageEinschraenkungen-typId']
['kantonId', 'auflageEinschraenkungen', 'auflageE

Index(['idPerson', 'name', 'vorname', 'geschlechtId', 'geburtsdatum',
       'jahrgang', 'sprachkompetenzIds', 'nationalitaet1Id',
       'nationalitaet2Id', 'gln', 'uid', 'adressenBewilligungen-0-kantonId',
       'adressenBewilligungen-0-praxisname', 'adressenBewilligungen-0-strasse',
       'adressenBewilligungen-0-plz', 'adressenBewilligungen-0-ort',
       'adressenBewilligungen-0-telefon1',
       'adressenBewilligungen-0-emailadresse',
       'adressenBewilligungen-0-zusatz1', 'adressenBewilligungen-0-zusatz2',
       'adressenBewilligungen-1-kantonId',
       'adressenBewilligungen-1-praxisname', 'adressenBewilligungen-1-strasse',
       'adressenBewilligungen-1-plz', 'adressenBewilligungen-1-ort',
       'adressenBewilligungen-1-telefon1',
       'adressenBewilligungen-1-emailadresse',
       'adressenBewilligungen-1-zusatz1', 'adressenBewilligungen-1-zusatz2',
       'adressenBewilligungen-2-kantonId',
       'adressenBewilligungen-2-praxisname', 'adressenBewilligungen-2-stra

In [23]:
pd.DataFrame(output,index=[0]).T

0
idPerson                                    11036
name                            Schweizer-Vüllers
vorname                                   Andreas
geschlechtId                                12000
geburtsdatum                                 None
...                                           ...
abschluesse-2-organisationName                   
abschluesse-2-organisationPlz                    
abschluesse-2-organisationOrt                    
canDoMutationsantrag                        False
canHaveBewilligungen                         True

[99 rows x 1 columns]

Dank zwei Sessions mit Plotti haben wir jetzt den Code soweit genagelt, damit er flache Daten erzeugt und dabei auch berücksichtigt, dass es pro Unterkategorie bis zu drei Einträge haben kann. 

In einem nächsten Schritt geht es darum die einzelnen Daten einzulesen und in einer Datei zu speichern. 